### PR TITLE
ARROW-14699: [C++] Fix lz4 undefined behaviour issues

### PR DESCRIPTION
**Pull request by pitrou**
LZ4 got a bunch of undefined behaviour fixes on its development branch, bump the bundled version.

### CODE REVIEW

**Comment by pitrou** [comment]
@github-actions crossbow submit fedora-r-clang-sanitizer

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g r

File: r/inst/build_arrow_static.sh [hunk 1]
```diff
@@ -82,5 +82,7 @@ ${CMAKE} -DARROW_BOOST_USE_SHARED=OFF \
     ${EXTRA_CMAKE_FLAGS} \
     -G ${CMAKE_GENERATOR:-"Unix Makefiles"} \
     ${SOURCE_DIR}
-${CMAKE} --build . --target install
+
+${CMAKE} --build . --parallel --target install
```
Comment (nealrichardson) [inline r/inst/build_arrow_static.sh:86]: Wait, so unless you add --parallel, it won't use the `-j` jobs set in MAKEFLAGS?

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-fedora-r-clang-sanitizer

File: r/inst/build_arrow_static.sh (hunk 1 above)
Comment (pitrou) [reply r/inst/build_arrow_static.sh:86]: Ok, it seems that wasn't necessary actually. MAKEFLAGS was simply not set correctly :-)

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g r

**Comment by pitrou** [comment]
@jonkeane Can you look at the test results above?

**Comment by jonkeane** [comment]
Those all look ok / failures that have other causes. There are a few builds which were simply stopped (e.g. [test-r-ubuntu-21.04](https://github.com/ursacomputing/crossbow/runs/4236526937?check_suite_focus=true#step:7:5332)), and then the other failures are related to DuckDB which will be addressed when https://github.com/apache/arrow/pull/11482 merges

**Pull request closed without merging**